### PR TITLE
feat(tti): increase robustness of perf metrics

### PR DESCRIPTION
**Pull request by patrickhulce**
First round of perf metric investigation, to be judged using #1936 :) Follow-up work includes adding in network reverse search methods for comparison.

- Adds two additional TTI candidates that don't rely on visual completeness.
	- First 500ms window after fMP where 90th percentile EIL is <50ms
	- First 5000ms window after fMP where 90th percentile EIL is <50ms (if at the end of trace, uses a minimum window of as low as 1000ms)
- Falls back to fMP when speedline fails to find screenshots in trace.
- Adds timing data to trace-of-tab computed artifact.

### CODE REVIEW

File: lighthouse-core/audits/time-to-interactive.js [hunk 1]
```diff
@@ -90,45 +187,23 @@ class TTIMetric extends Audit {
...
-        }
-        // Get our expected latency for the time window
-        const latencies = TracingProcessor.getRiskToResponsiveness(
-          model, trace, startTime, endTime, percentiles);
-        const estLatency = latencies[0].time;
-        foundLatencies.push({
-          estLatency: estLatency,
-          startTime: startTime.toFixed(1)
-        });
+      const times = {fmpTiming, visuallyReadyTiming, endOfTraceTime};
```
Comment (paulirish) [inline lighthouse-core/audits/time-to-interactive.js]: can you put a `lastTraceEvtTs` into trace-of-tab?

```js
const traceEndTsInMS = trace.traceEvents.reduce(
          (acc, traceEvent) => Math.max(acc, traceEvent.ts), 0) / 1000;
```
it's from will's https://github.com/GoogleChrome/lighthouse/pull/1936/files#diff-fd99fc8068c6555d1fb3d1c70df7c29cR82

wec can then reuse here (and maybe drop the traceviewer dep) and also in will's #1936

File: lighthouse-core/test/audits/estimated-input-latency-test.js [hunk 2]
```diff
@@ -37,8 +37,8 @@ describe('Performance: estimated-input-latency audit', () => {
     const artifacts = generateArtifactsWithTrace({traceEvents: pwaTrace});
     return Audit.audit(artifacts).then(output => {
       assert.equal(output.debugString, undefined);
-      assert.equal(output.rawValue, 17.4);
-      assert.equal(output.displayValue, '17.4ms');
+      assert.equal(output.rawValue, 16.7);
```
Comment (paulirish) [inline lighthouse-core/test/audits/estimated-input-latency-test.js]: to be clear, these changes are only due to the shift from speedline.first to FMP. yes?

File: lighthouse-core/gather/computed/trace-of-tab.js [hunk 3]
```diff
@@ -92,7 +92,16 @@ class TraceOfTab extends ComputedArtifact {
       .filter(e => e.pid === startedInPageEvt.pid)
       .sort((event0, event1) => event0.ts - event1.ts);
 
+    const metrics = {navigationStart, firstPaint, firstContentfulPaint, firstMeaningfulPaint};
+    const timings = {};
+
+    Object.keys(metrics).forEach(metric => {
+      timings[metric + 'Ts'] = metrics[metric] && metrics[metric].ts / 1000;
```
Comment (paulirish) [inline lighthouse-core/gather/computed/trace-of-tab.js]: the structure of the extInfo objects in TTI/FMP includes both a timings object and a timestamps object. should we do the same here?

File: lighthouse-core/gather/computed/trace-of-tab.js [hunk 4]
```diff
@@ -92,7 +92,16 @@ class TraceOfTab extends ComputedArtifact {
...
+      timings[metric] = timings[metric + 'Ts'] - timings.navigationStartTs;
```
Comment (paulirish) [inline lighthouse-core/gather/computed/trace-of-tab.js]: adding this guy in here means that if we pass along the data in extInfo then tools like pwmetrics could skip this work: https://github.com/paulirish/pwmetrics/blob/master/lib/metrics.ts#L54-L61

File: lighthouse-core/audits/time-to-interactive.js [hunk 5]
```diff
@@ -37,6 +37,99 @@ class TTIMetric extends Audit {
...
+  }
+
+  /**
+   * @param {{fmpTiming: number, visuallyReadyTiming: number, endOfTraceTime: number}} times
+   * @param {{model: !Object, trace: !Object}} data
+   * @return {{timeInMs: number|undefined, currentLatency: number, foundLatencies: !Array}}
+   */
+  static findTTIAlpha(times, data) {
+    return TTIMetric._forwardWindowTTI(
+      Math.max(times.fmpTiming, times.visuallyReadyTiming),
```
Comment (paulirish) [inline lighthouse-core/audits/time-to-interactive.js]: can you add a comment here to clarify you are using fmp as fallback?

File: lighthouse-core/audits/estimated-input-latency.js [hunk 6]
```diff
@@ -44,9 +44,9 @@ class EstimatedInputLatency extends Audit {
     };
   }
 
-  static calculate(speedline, model, trace) {
+  static calculate(tabTrace, model, trace) {
     // Use speedline's first paint as start of range for input latency check.
```
Comment (patrickhulce) [inline lighthouse-core/audits/estimated-input-latency.js]: done

File: lighthouse-core/test/audits/estimated-input-latency-test.js (hunk 2 above)
Comment (patrickhulce) [inline lighthouse-core/test/audits/estimated-input-latency-test.js]: yup

File: lighthouse-core/gather/computed/trace-of-tab.js (hunk 3 above)
Comment (patrickhulce) [inline lighthouse-core/gather/computed/trace-of-tab.js]: sgtm

File: lighthouse-core/audits/time-to-interactive.js (hunk 1 above)
Comment (patrickhulce) [inline lighthouse-core/audits/time-to-interactive.js]: done, but naming `traceEnd` :) to match with event names :)

File: lighthouse-core/audits/time-to-interactive.js (hunk 5 above)
Comment (patrickhulce) [inline lighthouse-core/audits/time-to-interactive.js]: done

File: lighthouse-core/audits/time-to-interactive.js [hunk 7]
```diff
@@ -63,9 +63,6 @@ class TTIMetric extends Audit {
         return {currentLatency, foundLatencies};
       }
 
-      // Clip to the trace end
-      endTime = Math.min(endTime, maxTime);
```
Comment (paulirish) [inline lighthouse-core/audits/time-to-interactive.js]: hmm! why'd you nuke this?
 
fwiw: https://docs.google.com/document/d/1b9slyaB9yho91YTOkAQfpCdULFkZM9LqsipcX3t7He8/edit#heading=h.ooalxriljq1i has details on the clipping approach for EIL

cc @brendankenny

Comment (patrickhulce) [inline lighthouse-core/audits/time-to-interactive.js]: oh I added this when I did the 5000ms - 1000ms if we reached the end of trace for WPT TTI, but after our convo on Friday about how fake that is I nuked the variable window and just kept the 5 seconds so it wasn't necessary anymore

Comment (paulirish) [inline lighthouse-core/audits/time-to-interactive.js]: gotcha. ok

**Comment by patrickhulce** [comment]
> want to update pwmetrics-events so we can get these new TTIs plotted in the trace?

done

image

**Pull request merged**